File: Programming Languages/pub/ch8/Excep.java
// javac Excep.java && java Excep

class ArgsException extends Exception {
    private String msg;
    public ArgsException(String msg) { this.msg=msg; }
    public String toString()         { return msg; }
}

public class Excep {

    static void foo(int i) throws ArgsException {
	if (i<1)
	    throw new ArgsException("too few args");
	// much more code
	System.err.println("foo() done");
    }

    public static void main(String[] args) {
	try {
	    foo(args.length);
	    // much more code
	    System.err.println("ok");
	} catch (ArgsException e) {
	    System.err.println("not ok: "+e);
	}
    }

}

File: Programming Languages/pub/ch8/Excep.cc
// g++ -Wall -o Excep Excep.cc && ./Excep

#include <iostream>
#include <string>

using namespace std;

class ArgsException {
  string msg;
public:
  ArgsException(string msg) : msg(msg) {}
  friend ostream& operator<<(ostream &os, const ArgsException& e) {
    return os << e.msg;
  }
};

class Excep {
public:
  static void foo(int i) throw(ArgsException) {
    if (i<2)
      throw ArgsException("too few args");
    // much more code
    cerr << "foo() done" << endl;
  }
};

int main(int argc, char *argv[]) {
  try {
    Excep::foo(argc);
    // much more code
    cerr << "ok" << endl;
  } catch (ArgsException e) {
    cerr << "not ok: " << e << endl;
  }
}
